sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/client/DefaultHttpClient.java

```diff
@@ -204,33 +216,28 @@ private HttpResponse receiveResponse(HttpRequest httpRequest, HttpURLConnection
         try {
             int responseCode = connection.getResponseCode();
 
-            Headers responseHeaders = new Headers();
+            Map<String, List<String>> hucHeaders = connection.getHeaderFields();
+            Headers responseHeaders = new Headers((int) (hucHeaders.size() / 0.75F));
 
             for (Map.Entry<String, List<String>> entry : connection.getHeaderFields().entrySet()) {
                 if (entry.getKey() != null) {
-                    for (String headerValue : entry.getValue()) {
-                        responseHeaders.add(HeaderName.fromString(entry.getKey()), headerValue);
-                    }
+                    responseHeaders.add(HeaderName.fromString(entry.getKey()), entry.getValue());
                 }
             }
 
-            ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+            AccessibleByteArrayOutputStream outputStream = new AccessibleByteArrayOutputStream();
 
             try (InputStream errorStream = connection.getErrorStream();
                  InputStream inputStream = (errorStream == null) ? connection.getInputStream() : errorStream) {
-                byte[] buffer = new byte[1024];
+                byte[] buffer = new byte[8192];
                 int length;
                 while ((length = inputStream.read(buffer)) != -1) {
                     outputStream.write(buffer, 0, length);
                 }
             }
 
-            return new DefaultHttpClientResponse(
-                httpRequest,
-                responseCode,
-                responseHeaders,
-                BinaryData.fromByteBuffer(ByteBuffer.wrap(outputStream.toByteArray()))
-            );
+            return new DefaultHttpClientResponse(httpRequest, responseCode, responseHeaders,
+                BinaryData.fromByteBuffer(outputStream.toByteBuffer()));
```

Comment:
[non-blocking] we can go even further and don't buffer the output stream (like https://github.com/Azure/azure-sdk-for-java/pull/38206/files#diff-37fd5025949a7a762faf4a100376503f75a80fe98f5cd5838e13cdc28e9c0886) 

caller is responsible for closing it anyway. Also, this is where closeable BinaryData would come really handy